synapse/federation/transport/server/__init__.py

```diff
@@ -286,7 +288,50 @@ async def on_GET(
                 },
             )
 
-        return 200, {"sub": user_id}
+        user_id, userinfo_fields = token_info
+        userinfo: JsonDict = {"sub": user_id}
+
+        if userinfo_fields:
+            if OpenIdUserInfoFields.DISPLAY_NAME in userinfo_fields:
+                localpart = UserID.from_string(user_id).localpart
+                userinfo[
+                    OpenIdUserInfoFields.DISPLAY_NAME
+                ] = await self.hs.get_datastores().main.get_profile_displayname(
+                    localpart
+                )
+            if OpenIdUserInfoFields.AVATAR_URL in userinfo_fields:
+                localpart = UserID.from_string(user_id).localpart
+                userinfo[
+                    OpenIdUserInfoFields.AVATAR_URL
+                ] = await self.hs.get_datastores().main.get_profile_avatar_url(
+                    localpart
+                )
+            if OpenIdUserInfoFields.ROOM_POWERLEVELS in userinfo_fields:
+                userinfo[
+                    OpenIdUserInfoFields.ROOM_POWERLEVELS
+                ] = await self._get_powerlevels(user_id)
+
+        return 200, userinfo
+
+    async def _get_powerlevels(self, user_id: str) -> Dict[str, Dict[str, Any]]:
+        room_list = await self.hs.get_datastores().main.get_rooms_for_local_user_where_membership_is(
+            user_id=user_id, membership_list=[Membership.JOIN]
+        )
+
+        power_levels = {}
+        for room in room_list:
+            room_state = await self.hs.get_state_handler().get_current_state(
+                room.room_id
+            )
+            room_power_levels = room_state.get((EventTypes.PowerLevels, ""))
+            if room_power_levels is not None:
+                stripped_content = copy.deepcopy(room_power_levels.content)
+                stripped_content["users"] = {
+                    k: v for k, v in stripped_content["users"].items() if k == user_id
+                }
```

Comment:
a comment explaining what we're doing here, with an example, would be helpful. It's not entirely clear at first glance.